core/src/main/kotlin/xtdb/trie/HashTrie.kt

```diff
@@ -97,11 +101,13 @@ fun toMergePlan(segments: List<ISegment>, pathPred: Predicate<ByteArray>?): List
             mpNodes.any { it.node is RecencyBranch } -> {
                 val newMpNodes = mutableListOf<MergePlanNode>()
                 for (mpNode in mergePlanTask.mpNodes) {
-                    // TODO filter by recencies #3166
                     val recencies = mpNode.node.recencies
                     if (recencies != null) {
-                        for (i in recencies.indices) {
+                        var i = 0
+                        while (i < recencies.size && recencies[i] <= minRecency) i++
+                        while (i < recencies.size) {
                             newMpNodes += MergePlanNode(mpNode.segment, mpNode.node.recencyNode(i))
+                            i++
```

Comment:
This 5 lines are actually the logic, so probably good to have a second :eyes: over this.